File: implementations/typescript/packages/llmproxy/src/openai-chat.ts
// SPDX-License-Identifier: Apache-2.0
import { request as httpsRequest } from "node:https";
import { canonicalJson, parseJson, record } from "@psp-cdl/core";
import { aggregateCapabilities } from "@psp-cdl/cdl";
import { bounded } from "@psp-cdl/api-server/persistence";
import type { CapabilitySource } from "@psp-cdl/mcpproxy";
import type { ProviderRegistration } from "./loop.js";

export const OPENAI_CHAT_PROFILE = "PSP-OPENAI-CHAT-0.1";
export const OPENAI_CHAT_MODEL = "gpt-4.1-mini-2025-04-14";
export const OPENAI_CHAT_REVISION = "chat-v1-gpt-4.1-mini-2025-04-14-psp-0.1";
export const OPENAI_CHAT_INPUT_RESERVATION = 1_047_576;
const MAX_BYTES = 1_048_576;
const ERROR_CODES = new Set(["INVALID_CONFIGURATION", "INVALID_REQUEST", "INVALID_RESPONSE", "REQUEST_TOO_LARGE", "RESPONSE_TOO_LARGE",
  "BUDGET_EXHAUSTED", "PROVIDER_BUSY", "CANCELLED", "DEADLINE_EXCEEDED", "PROVIDER_HTTP_ERROR", "PROVIDER_FAILED", "HOST_ERROR"]);
export interface OpenAIChatLimits {
  maxRequestBytes: number; maxResponseBytes: number; maxOutputTokens: number;
  maxCalls: number; budgetTokens: number; timeoutMs: number;
}
export interface OpenAIChatReply { status: number; contentType: string; body: Uint8Array }
/** Trusted synthetic callback only. This callback receives no API credentials. */
export type OfflineChatTransport = (body: Uint8Array, signal: AbortSignal) => OpenAIChatReply | Promise<OpenAIChatReply>;
interface CommonConfig { sources: CapabilitySource[]; complete: true; now: () => number; limits: OpenAIChatLimits }
export type OpenAIChatConfig = CommonConfig & (
  { mode: "live"; allowLive: true; apiKey: string } |
  { mode: "offline"; transport: OfflineChatTransport }
);
export class ProviderError extends Error {
  constructor(public readonly code: string) { super(code); this.name = "ProviderError"; }
}
const fail = (code: string): never => { throw new ProviderError(code); };
const integer = (v: unknown): v is number => Number.isSafeInteger(v) && (v as number) >= 0;
const exact = (v: unknown, keys: string[]): v is Record<string, any> => record(v) && Object.keys(v).sort().join(",") === keys.sort().join(",");
const allowed = (v: Record<string, unknown>, keys: string[]): boolean => Object.keys(v).every(k => keys.includes(k));
function copy(v: unknown, code: string): any { try { return bounded(v); } catch { return fail(code); } }

function encodeRequest(input: unknown, limits: OpenAIChatLimits): { body: Uint8Array; names: string[] } {
  const value = copy(input, "INVALID_REQUEST");
  if (!exact(value, ["messages", "tools"]) || !Array.isArray(value.messages) || !Array.isArray(value.tools) ||
      value.messages.length < 2 || value.tools.length > 128) fail("INVALID_REQUEST");
  const tools = [...value.tools].sort((a, b) => String(a?.name) < String(b?.name) ? -1 : String(a?.name) > String(b?.name) ? 1 : 0);
  const names: string[] = [];
  const wireTools = tools.map((t, i) => {
    if (!exact(t, ["name", "inputSchema", "outputSchema"]) || typeof t.name !== "string" ||
        !/^[A-Za-z0-9_.-]{1,256}$/.test(t.name) || names.includes(t.name) || !record(t.inputSchema) || !record(t.outputSchema)) fail("INVALID_REQUEST");
    names.push(t.name);
    return { type: "function", function: { name: `psp_tool_${i}`, description: t.name, parameters: t.inputSchema } };
  });
  const messages: Record<string, unknown>[] = [];
  let pending: { name: string; id: string } | undefined, call = 0;
  for (const [i, m] of value.messages.entries()) {
    if (!record(m)) fail("INVALID_REQUEST");
    if (i < 2) {
      if (!exact(m, ["role", "content"]) || m.role !== (i === 0 ? "system" : "user") || typeof m.content !== "string") fail("INVALID_REQUEST");
      messages.push(m);
    } else if (!pending && exact(m, ["role", "call"]) && m.role === "assistant") {
      if (!exact(m.call, ["name", "arguments"]) || !names.includes(m.call.name) || !record(m.call.arguments)) fail("INVALID_REQUEST");
      const id = `psp_call_${call++}`;
      pending = { name: m.call.name, id };
      messages.push({ role: "assistant", content: null, tool_calls: [{ id, type: "function", function: {
        name: `psp_tool_${names.indexOf(m.call.name)}`, arguments: canonicalJson(m.call.arguments)
      } }] });
    } else if (pending && exact(m, ["role", "name", "data"]) && m.role === "tool" && m.name === pending.name && record(m.data)) {
      messages.push({ role: "tool", tool_call_id: pending.id, content: canonicalJson(m.data) });
      pending = undefined;
    } else fail("INVALID_REQUEST");
  }
  if (pending) fail("INVALID_REQUEST");
  const request = { model: OPENAI_CHAT_MODEL, messages, stream: false, n: 1, store: false,
    max_completion_tokens: limits.maxOutputTokens,
    ...(tools.length ? { tools: wireTools, parallel_tool_calls: false, tool_choice: "auto" } : {}) };
  let bytes: Uint8Array;
  try { bytes = Buffer.from(canonicalJson(request), "utf8"); } catch { return fail("REQUEST_TOO_LARGE"); }
  if (bytes.length > limits.maxRequestBytes) fail("REQUEST_TOO_LARGE");
  return { body: bytes, names };
}

function decodeReply(reply: OpenAIChatReply, names: string[], limits: OpenAIChatLimits): Record<string, unknown> {
  if (!exact(reply, ["status", "contentType", "body"]) || !integer(reply.status)) fail("INVALID_RESPONSE");
  if (reply.status !== 200) fail("PROVIDER_HTTP_ERROR");
  if (typeof reply.contentType !== "string" || !/^application\/json(?:\s*;\s*charset=utf-8)?$/i.test(reply.contentType)) fail("INVALID_RESPONSE");
  if (!(reply.body instanceof Uint8Array)) fail("INVALID_RESPONSE");
  if (reply.body.length > limits.maxResponseBytes) fail("RESPONSE_TOO_LARGE");
  let r: any;
  try { r = parseJson(new TextDecoder("utf-8", { fatal: true, ignoreBOM: true }).decode(reply.body)); } catch { return fail("INVALID_RESPONSE"); }
  if (!record(r) || !allowed(r, ["id", "object", "created", "model", "choices", "usage", "system_fingerprint", "service_tier"]) ||
      typeof r.id !== "string" || !r.id.length || !integer(r.created) || r.object !== "chat.completion" || r.model !== OPENAI_CHAT_MODEL ||
      !Array.isArray(r.choices) || r.choices.length !== 1) fail("INVALID_RESPONSE");
  const usage: any = r.usage;
  if (!record(usage) || !allowed(usage, ["prompt_tokens", "completion_tokens", "total_tokens", "prompt_tokens_details", "completion_tokens_details"]) ||
      !integer(usage.prompt_tokens) || !integer(usage.completion_tokens) || !integer(usage.total_tokens) ||
      usage.prompt_tokens > OPENAI_CHAT_INPUT_RESERVATION || usage.completion_tokens > limits.maxOutputTokens ||
      usage.total_tokens !== usage.prompt_tokens + usage.completion_tokens) fail("INVALID_RESPONSE");
  const c: any = r.choices[0];
  if (!record(c) || !allowed(c, ["index", "message", "finish_reason", "logprobs"]) || c.index !== 0 || c.logprobs != null) fail("INVALID_RESPONSE");
  const m: any = c.message;
  if (!record(m) || !allowed(m, ["role", "content", "refusal", "annotations", "tool_calls"]) || m.role !== "assistant" || m.refusal != null ||
      m.annotations !== undefined && (!Array.isArray(m.annotations) || m.annotations.length !== 0)) fail("INVALID_RESPONSE");
  if (c.finish_reason === "stop" && typeof m.content === "string" && (m.tool_calls === undefined || Array.isArray(m.tool_calls) && m.tool_calls.length === 0)) {
    return copy({ type: "final", text: m.content }, "INVALID_RESPONSE");
  }
  if (c.finish_reason !== "tool_calls" || !(m.content === null || m.content === "") || !Array.isArray(m.tool_calls) || m.tool_calls.length !== 1) fail("INVALID_RESPONSE");
  const t = m.tool_calls[0];
  if (!exact(t, ["id", "type", "function"]) || typeof t.id !== "string" || !/^[A-Za-z0-9_-]{1,128}$/.test(t.id) ||
      t.type !== "function" || !exact(t.function, ["name", "arguments"]) || typeof t.function.arguments !== "string") fail("INVALID_RESPONSE");
  const index = names.findIndex((_, i) => `psp_tool_${i}` === t.function.name);
  if (index < 0) fail("INVALID_RESPONSE");
  let args: unknown;
  try { args = parseJson(t.function.arguments); } catch { return fail("INVALID_RESPONSE"); }
  if (!record(args)) fail("INVALID_RESPONSE");
  return copy({ type: "tool", name: names[index], arguments: args }, "INVALID_RESPONSE");
}

/** Fixed TLS endpoint, no redirects, retries, proxy lookup, decompression or SDK. */
function liveTransport(key: string, maxBytes: number, body: Uint8Array, signal: AbortSignal): Promise<OpenAIChatReply> {
  return new Promise((resolve, reject) => {
    const req = httpsRequest("https://api.openai.com/v1/chat/completions", {
      method: "POST", agent: false, signal, maxHeaderSize: 16_384, rejectUnauthorized: true,
      headers: { Authorization: `Bearer ${key}`, "Content-Type": "application/json", Accept: "application/json",
        "Accept-Encoding": "identity", "Content-Length": body.length }
    }, res => {
      const rejectResponse = (code: string) => { reject(new ProviderError(code)); res.destroy(); req.destroy(); };
      if (res.statusCode !== 200) { rejectResponse("PROVIDER_HTTP_ERROR"); return; }
      if (res.headers["content-encoding"] && res.headers["content-encoding"] !== "identity") { rejectResponse("INVALID_RESPONSE"); return; }
      const length = res.headers["content-length"];
      if (length !== undefined && (!/^\d+$/.test(length) || Number(length) > maxBytes)) { rejectResponse("RESPONSE_TOO_LARGE"); return; }
      const chunks: Buffer[] = []; let size = 0;
      res.on("data", (chunk: Buffer) => {
        size += chunk.length;
        if (size > maxBytes) rejectResponse("RESPONSE_TOO_LARGE"); else chunks.push(chunk);
      });
      res.on("error", () => reject(new ProviderError("PROVIDER_FAILED")));
      res.on("end", () => {
        if (!res.complete) { rejectResponse("INVALID_RESPONSE"); return; }
        resolve({ status: 200, contentType: res.headers["content-type"] ?? "", body: Buffer.concat(chunks) });
      });
    });
    req.on("error", () => reject(new ProviderError("PROVIDER_FAILED")));
    req.end(body);
  });
}

/** Host-owned closure. Construction and import never perform network I/O. */
export function createOpenAIChatProvider(config: OpenAIChatConfig): ProviderRegistration {
  if (!record(config) || !["live", "offline"].includes(config.mode) || typeof config.now !== "function") fail("INVALID_CONFIGURATION");
  const live = config.mode === "live";
  if (!exact(config, ["mode", "sources", "complete", "now", "limits", ...(live ? ["allowLive", "apiKey"] : ["transport"])])) fail("INVALID_CONFIGURATION");
  if (live && (config.allowLive !== true || typeof config.apiKey !== "string" || !/^[\x21-\x7e]{1,4096}$/.test(config.apiKey)) ||
      !live && typeof config.transport !== "function") fail("INVALID_CONFIGURATION");
  const limits: OpenAIChatLimits = copy(config.limits, "INVALID_CONFIGURATION"), sources = copy(config.sources, "INVALID_CONFIGURATION");
  if (!exact(limits, ["maxRequestBytes", "maxResponseBytes", "maxOutputTokens", "maxCalls", "budgetTokens", "timeoutMs"]) ||
      Object.values(limits).some(v => !integer(v) || v < 1) || limits.maxRequestBytes > MAX_BYTES || limits.maxResponseBytes > MAX_BYTES ||
      limits.maxOutputTokens > 32_768 || limits.maxCalls > 32 || limits.timeoutMs > 120_000) fail("INVALID_CONFIGURATION");
  try { aggregateCapabilities(sources, config.complete); } catch { fail("INVALID_CONFIGURATION"); }
  if (config.complete !== true) fail("INVALID_CONFIGURATION");
  const now = config.now, key = live ? config.apiKey : "", transport = live ? undefined : config.transport;
  let calls = 0, reserved = 0, busy = false;
  return { id: live ? "openai-chat" : "openai-chat-offline", revision: OPENAI_CHAT_REVISION, sources, complete: true,
    invoke: async (input, options) => {
      if (busy) fail("PROVIDER_BUSY");
      busy = true;
      let timer: ReturnType<typeof setInterval> | undefined;
      let operation: Promise<OpenAIChatReply> | undefined;
      const controller = new AbortController();
      try {
        if (!exact(options, ["deadline", "cancelled"]) || !integer(options.deadline) || typeof options.cancelled !== "function") fail("INVALID_REQUEST");
        const controls = { deadline: options.deadline, cancelled: options.cancelled }, start = performance.now();
        const check = () => {
          let time: number, cancelled: boolean;
          try { time = now(); cancelled = controls.cancelled(); } catch { return fail("HOST_ERROR"); }
          if (!integer(time) || typeof cancelled !== "boolean") fail("HOST_ERROR");
          if (cancelled) fail("CANCELLED");
          if (time >= controls.deadline || performance.now() - start >= limits.timeoutMs) fail("DEADLINE_EXCEEDED");
        };
        check();
        const { body, names } = encodeRequest(input, limits);
        check();
        const reservation = OPENAI_CHAT_INPUT_RESERVATION + limits.maxOutputTokens;
        if (calls >= limits.maxCalls || reservation > limits.budgetTokens - reserved) fail("BUDGET_EXHAUSTED");
        calls++; reserved += reservation;
        const interrupted = new Promise<never>((_, reject) => {
          timer = setInterval(() => { try { check(); } catch (e) { reject(e); controller.abort(); } }, 10);
        });
        operation = Promise.resolve().then(() => {
          check();
          return live ? liveTransport(key, limits.maxResponseBytes, body, controller.signal) : transport!(body, controller.signal);
        });
        const reply = await Promise.race([operation, interrupted]);
        check();
        const result = decodeReply(reply, names, limits);
        check();
        return result;
      } catch (e) {
        // Never retain a backend exception, request, response, credential or cause.
        throw new ProviderError(e instanceof ProviderError && ERROR_CODES.has(e.code) ? e.code : "PROVIDER_FAILED");
      } finally {
        if (timer !== undefined) clearInterval(timer);
        controller.abort();
        // A cancelled but uncooperative transport still owns the pending slot.
        if (operation) void operation.then(() => { busy = false; }, () => { busy = false; });
        else busy = false;
      }
    }
  };
}

File: implementations/python/packages/llmproxy/src/psp_cdl_llmproxy/openai_chat.py
# SPDX-License-Identifier: Apache-2.0
"""Opt-in, buffered OpenAI Chat mapping. No import-time network or credential lookup."""
import http.client
import re
import socket
import ssl
import threading
import time

from psp_cdl_core import canonical_json, parse_json
from psp_cdl_cdl import aggregate_capabilities
from psp_cdl_api_server.persistence import bounded, integer

OPENAI_CHAT_PROFILE = "PSP-OPENAI-CHAT-0.1"
OPENAI_CHAT_MODEL = "gpt-4.1-mini-2025-04-14"
OPENAI_CHAT_REVISION = "chat-v1-gpt-4.1-mini-2025-04-14-psp-0.1"
OPENAI_CHAT_INPUT_RESERVATION = 1_047_576
MAX_BYTES = 1_048_576
ERROR_CODES = frozenset(("INVALID_CONFIGURATION", "INVALID_REQUEST", "INVALID_RESPONSE", "REQUEST_TOO_LARGE", "RESPONSE_TOO_LARGE",
                        "BUDGET_EXHAUSTED", "PROVIDER_BUSY", "CANCELLED", "DEADLINE_EXCEEDED", "PROVIDER_HTTP_ERROR", "PROVIDER_FAILED", "HOST_ERROR"))


class ProviderError(ValueError):
    def __init__(self, code):
        super().__init__(code)
        self.code = code


def _fail(code):
    raise ProviderError(code)


def _exact(value, keys):
    return type(value) is dict and set(value) == set(keys)


def _allowed(value, keys):
    return type(value) is dict and set(value) <= set(keys)


def _copy(value, code):
    try:
        return bounded(value)
    except Exception:
        pass
    _fail(code)


def _parse(source):
    try:
        return parse_json(source)
    except Exception:
        pass
    _fail("INVALID_RESPONSE")


def _encode_request(value, limits):
    value = _copy(value, "INVALID_REQUEST")
    if (not _exact(value, ["messages", "tools"]) or type(value["messages"]) is not list
            or type(value["tools"]) is not list or len(value["messages"]) < 2 or len(value["tools"]) > 128):
        _fail("INVALID_REQUEST")
    # Validate names before sorting so malformed values fail identically in both languages.
    for tool in value["tools"]:
        if (not _exact(tool, ["name", "inputSchema", "outputSchema"]) or type(tool["name"]) is not str
                or not re.fullmatch(r"[A-Za-z0-9_.-]{1,256}", tool["name"])
                or type(tool["inputSchema"]) is not dict or type(tool["outputSchema"]) is not dict):
            _fail("INVALID_REQUEST")
    tools = sorted(value["tools"], key=lambda t: t["name"])
    names = [t["name"] for t in tools]
    if len(set(names)) != len(names):
        _fail("INVALID_REQUEST")
    wire_tools = [{"type": "function", "function": {"name": f"psp_tool_{i}", "description": t["name"], "parameters": t["inputSchema"]}} for i, t in enumerate(tools)]
    messages, pending, call = [], None, 0
    for i, message in enumerate(value["messages"]):
        if type(message) is not dict:
            _fail("INVALID_REQUEST")
        if i < 2:
            if (not _exact(message, ["role", "content"]) or message["role"] != ("system" if i == 0 else "user")
                    or type(message["content"]) is not str):
                _fail("INVALID_REQUEST")
            messages.append(message)
        elif pending is None and _exact(message, ["role", "call"]) and message["role"] == "assistant":
            item = message["call"]
            if not _exact(item, ["name", "arguments"]) or item["name"] not in names or type(item["arguments"]) is not dict:
                _fail("INVALID_REQUEST")
            call_id = f"psp_call_{call}"
            call += 1
            pending = (item["name"], call_id)
            messages.append({"role": "assistant", "content": None, "tool_calls": [{"id": call_id, "type": "function", "function": {
                "name": f"psp_tool_{names.index(item['name'])}", "arguments": canonical_json(item["arguments"])
            }}]})
        elif (pending is not None and _exact(message, ["role", "name", "data"]) and message["role"] == "tool"
              and message["name"] == pending[0] and type(message["data"]) is dict):
            messages.append({"role": "tool", "tool_call_id": pending[1], "content": canonical_json(message["data"])})
            pending = None
        else:
            _fail("INVALID_REQUEST")
    if pending is not None:
        _fail("INVALID_REQUEST")
    request = {"model": OPENAI_CHAT_MODEL, "messages": messages, "stream": False, "n": 1, "store": False,
               "max_completion_tokens": limits["maxOutputTokens"]}
    if tools:
        request.update(tools=wire_tools, parallel_tool_calls=False, tool_choice="auto")
    body = None
    try:
        body = canonical_json(request).encode("utf-8")
    except Exception:
        pass
    if body is None or len(body) > limits["maxRequestBytes"]:
        _fail("REQUEST_TOO_LARGE")
    return body, names


def _decode_reply(reply, names, limits):
    if not _exact(reply, ["status", "contentType", "body"]) or not integer(reply["status"]):
        _fail("INVALID_RESPONSE")
    if reply["status"] != 200:
        _fail("PROVIDER_HTTP_ERROR")
    if (type(reply["contentType"]) is not str or not re.fullmatch(r"application/json(?:\s*;\s*charset=utf-8)?", reply["contentType"], re.I)
            or type(reply["body"]) is not bytes):
        _fail("INVALID_RESPONSE")
    if len(reply["body"]) > limits["maxResponseBytes"]:
        _fail("RESPONSE_TOO_LARGE")
    source = None
    try:
        source = reply["body"].decode("utf-8", errors="strict")
    except Exception:
        pass
    if source is None:
        _fail("INVALID_RESPONSE")
    result = _parse(source)
    if (not _allowed(result, ["id", "object", "created", "model", "choices", "usage", "system_fingerprint", "service_tier"])
            or type(result.get("id")) is not str or not result["id"] or not integer(result.get("created"))
            or result.get("object") != "chat.completion" or result.get("model") != OPENAI_CHAT_MODEL
            or type(result.get("choices")) is not list or len(result["choices"]) != 1):
        _fail("INVALID_RESPONSE")
    usage = result.get("usage")
    if (not _allowed(usage, ["prompt_tokens", "completion_tokens", "total_tokens", "prompt_tokens_details", "completion_tokens_details"])
            or any(not integer(usage.get(k)) for k in ("prompt_tokens", "completion_tokens", "total_tokens"))
            or usage["prompt_tokens"] > OPENAI_CHAT_INPUT_RESERVATION or usage["completion_tokens"] > limits["maxOutputTokens"]
            or usage["total_tokens"] != usage["prompt_tokens"] + usage["completion_tokens"]):
        _fail("INVALID_RESPONSE")
    choice = result["choices"][0]
    if (not _allowed(choice, ["index", "message", "finish_reason", "logprobs"]) or not integer(choice.get("index"))
            or choice["index"] != 0 or choice.get("logprobs") is not None):
        _fail("INVALID_RESPONSE")
    message = choice.get("message")
    if (not _allowed(message, ["role", "content", "refusal", "annotations", "tool_calls"]) or message.get("role") != "assistant"
            or message.get("refusal") is not None or "annotations" in message and message["annotations"] != []):
        _fail("INVALID_RESPONSE")
    if (choice.get("finish_reason") == "stop" and type(message.get("content")) is str
            and ("tool_calls" not in message or message["tool_calls"] == [])):
        return _copy({"type": "final", "text": message["content"]}, "INVALID_RESPONSE")
    if (choice.get("finish_reason") != "tool_calls" or "content" not in message or message["content"] not in (None, "")
            or type(message.get("tool_calls")) is not list or len(message["tool_calls"]) != 1):
        _fail("INVALID_RESPONSE")
    tool = message["tool_calls"][0]
    if (not _exact(tool, ["id", "type", "function"]) or type(tool["id"]) is not str
            or not re.fullmatch(r"[A-Za-z0-9_-]{1,128}", tool["id"]) or tool["type"] != "function"
            or not _exact(tool["function"], ["name", "arguments"]) or type(tool["function"]["arguments"]) is not str):
        _fail("INVALID_RESPONSE")
    aliases = [f"psp_tool_{i}" for i in range(len(names))]
    if tool["function"]["name"] not in aliases:
        _fail("INVALID_RESPONSE")
    args = _parse(tool["function"]["arguments"])
    if type(args) is not dict:
        _fail("INVALID_RESPONSE")
    return _copy({"type": "tool", "name": names[aliases.index(tool["function"]["name"])], "arguments": args}, "INVALID_RESPONSE")


class _Call:
    def __init__(self):
        self.stop = threading.Event()
        self.connection = None
        self.socket = None

    def abort(self):
        self.stop.set()
        sock = self.socket
        # Shutdown interrupts reads; close alone may not interrupt a blocking read.
        if sock is not None:
            try:
                sock.shutdown(socket.SHUT_RDWR)
            except OSError:
                pass


def _live_transport(key, max_bytes, timeout, body, call):
    connection = http.client.HTTPSConnection("api.openai.com", timeout=timeout, context=ssl.create_default_context())
    call.connection = connection
    try:
        if call.stop.is_set():
            _fail("CANCELLED")
        connection.connect()
        call.socket = connection.sock
        if call.stop.is_set():
            _fail("CANCELLED")
        connection.request("POST", "/v1/chat/completions", body, headers={
            "Authorization": "Bearer " + key, "Content-Type": "application/json", "Accept": "application/json",
            "Accept-Encoding": "identity", "Content-Length": str(len(body)), "Connection": "close"
        })
        response = connection.getresponse()
        if response.status != 200:
            _fail("PROVIDER_HTTP_ERROR")
        if response.getheader("Content-Encoding", "identity") != "identity":
            _fail("INVALID_RESPONSE")
        length = response.getheader("Content-Length")
        if length is not None and (not re.fullmatch(r"[0-9]+", length) or int(length) > max_bytes):
            _fail("RESPONSE_TOO_LARGE")
        parts, size = [], 0
        while True:
            if call.stop.is_set():
                _fail("CANCELLED")
            chunk = response.read1(min(16_384, max_bytes + 1 - size))
            if not chunk:
                break
            size += len(chunk)
            if size > max_bytes:
                _fail("RESPONSE_TOO_LARGE")
            parts.append(chunk)
        if length is not None and size != int(length):
            _fail("PROVIDER_FAILED")
        return {"status": 200, "contentType": response.getheader("Content-Type", ""), "body": b"".join(parts)}
    finally:
        connection.close()


def create_openai_chat_provider(config):
    """Return the existing provider registration; credentials/budget remain host-owned."""
    if type(config) is not dict or config.get("mode") not in ("live", "offline") or not callable(config.get("now")):
        _fail("INVALID_CONFIGURATION")
    live = config["mode"] == "live"
    keys = ["mode", "sources", "complete", "now", "limits"] + (["allowLive", "apiKey"] if live else ["transport"])
    if not _exact(config, keys):
        _fail("INVALID_CONFIGURATION")
    if (live and (config["allowLive"] is not True or type(config["apiKey"]) is not str or not re.fullmatch(r"[\x21-\x7e]{1,4096}", config["apiKey"]))
            or not live and not callable(config["transport"])):
        _fail("INVALID_CONFIGURATION")
    limits, sources = _copy(config["limits"], "INVALID_CONFIGURATION"), _copy(config["sources"], "INVALID_CONFIGURATION")
    if (not _exact(limits, ["maxRequestBytes", "maxResponseBytes", "maxOutputTokens", "maxCalls", "budgetTokens", "timeoutMs"])
            or any(not integer(v) or v < 1 for v in limits.values()) or limits["maxRequestBytes"] > MAX_BYTES
            or limits["maxResponseBytes"] > MAX_BYTES or limits["maxOutputTokens"] > 32_768 or limits["maxCalls"] > 32
            or limits["timeoutMs"] > 120_000):
        _fail("INVALID_CONFIGURATION")
    valid_caps = False
    try:
        aggregate_capabilities(sources, config["complete"])
        valid_caps = config["complete"] is True
    except Exception:
        pass
    if not valid_caps:
        _fail("INVALID_CONFIGURATION")
    # Integer-valued JSON numbers have the same semantics as JS safe integers.
    limits = {k: int(v) for k, v in limits.items()}
    now, key, transport = config["now"], config.get("apiKey", ""), config.get("transport")
    lock, calls, reserved = threading.Lock(), 0, 0

    def invoke(value, options):
        nonlocal calls, reserved
        if not lock.acquire(blocking=False):
            _fail("PROVIDER_BUSY")
        call = _Call()
        code = None
        lifecycle = threading.Lock()
        state = {"started": False, "finished": False, "returned": False}
        try:
            if not _exact(options, ["deadline", "cancelled"]) or not integer(options["deadline"]) or not callable(options["cancelled"]):
                _fail("INVALID_REQUEST")
            deadline, cancelled, start = options["deadline"], options["cancelled"], time.monotonic()

            def check():
                valid = False
                try:
                    current, stopped = now(), cancelled()
                    valid = integer(current) and type(stopped) is bool
                except Exception:
                    pass
                if not valid:
                    _fail("HOST_ERROR")
                if stopped:
                    _fail("CANCELLED")
                if current >= deadline or (time.monotonic() - start) * 1000 >= limits["timeoutMs"]:
                    _fail("DEADLINE_EXCEEDED")

            check()
            body, names = _encode_request(value, limits)
            check()
            reservation = OPENAI_CHAT_INPUT_RESERVATION + limits["maxOutputTokens"]
            if calls >= limits["maxCalls"] or reservation > limits["budgetTokens"] - reserved:
                _fail("BUDGET_EXHAUSTED")
            calls += 1
            reserved += reservation
            done, outcome = threading.Event(), {}

            def worker():
                try:
                    if call.stop.is_set():
                        _fail("CANCELLED")
                    outcome["reply"] = (_live_transport(key, limits["maxResponseBytes"], limits["timeoutMs"] / 1000, body, call)
                                        if live else transport(body, call.stop))
                except Exception as exc:
                    outcome["error"] = exc.code if isinstance(exc, ProviderError) and exc.code in ERROR_CODES else "PROVIDER_FAILED"
                finally:
                    done.set()
                    with lifecycle:
                        state["finished"] = True
                        if state["returned"]:
                            lock.release()

            worker_thread = threading.Thread(target=worker, daemon=True, name="psp-openai-io")
            worker_thread.start()
            state["started"] = True
            while not done.wait(0.01):
                check()
            check()
            if "error" in outcome:
                _fail(outcome["error"])
            result = _decode_reply(outcome["reply"], names, limits)
            check()
            return result
        except Exception as exc:
            code = exc.code if isinstance(exc, ProviderError) and exc.code in ERROR_CODES else "PROVIDER_FAILED"
        finally:
            call.abort()
            with lifecycle:
                state["returned"] = True
                if not state["started"] or state["finished"]:
                    lock.release()
        # Raise outside the handler so backend exceptions are not retained as context.
        _fail(code)

    return {"id": "openai-chat" if live else "openai-chat-offline", "revision": OPENAI_CHAT_REVISION,
            "sources": sources, "complete": True, "invoke": invoke}
